# Classical Machine Learning

### Introduction

This notebook explores the [Iris](https://archive.ics.uci.edu/dataset/53/iris) dataset using classical machine learning methods such as neural networks, k-NN, etc. to perform classification. The purpose of this is to establish a baseline for our attempt at applying quantum machine learning in later stages.

### Expectation

... We expect high accuracy...

In [1]:
# Turn off warnings in Jupyter Notebook
import warnings
warnings.filterwarnings('ignore')

## Dataset

The dataset is publicly available and I have prepared a in the `datasets` directory to automate the process of downloading it, take a look at the `README.md` in that directory for further information. 

### Importing data

We use `polars` here instead of the well-known `pandas` library but they are basically interchangeable.

In [2]:
import polars as pl

DATASET_PATH = "../datasets/iris.csv"

data = pl.read_csv(DATASET_PATH)

### Exploratory Data Analysis (EDA)

We do some preliminary checks to get a sense of what we are working with.

In [3]:
data.head()

shape: (5, 5)
┌──────────────┬─────────────┬──────────────┬─────────────┬─────────────┐
│ sepal_length ┆ sepal_width ┆ petal_length ┆ petal_width ┆ class       │
│ ---          ┆ ---         ┆ ---          ┆ ---         ┆ ---         │
│ f64          ┆ f64         ┆ f64          ┆ f64         ┆ str         │
╞══════════════╪═════════════╪══════════════╪═════════════╪═════════════╡
│ 5.1          ┆ 3.5         ┆ 1.4          ┆ 0.2         ┆ Iris-setosa │
│ 4.9          ┆ 3.0         ┆ 1.4          ┆ 0.2         ┆ Iris-setosa │
│ 4.7          ┆ 3.2         ┆ 1.3          ┆ 0.2         ┆ Iris-setosa │
│ 4.6          ┆ 3.1         ┆ 1.5          ┆ 0.2         ┆ Iris-setosa │
│ 5.0          ┆ 3.6         ┆ 1.4          ┆ 0.2         ┆ Iris-setosa │
└──────────────┴─────────────┴──────────────┴─────────────┴─────────────┘

In [4]:
data.describe()

shape: (9, 6)
┌────────────┬──────────────┬─────────────┬──────────────┬─────────────┬────────────────┐
│ statistic  ┆ sepal_length ┆ sepal_width ┆ petal_length ┆ petal_width ┆ class          │
│ ---        ┆ ---          ┆ ---         ┆ ---          ┆ ---         ┆ ---            │
│ str        ┆ f64          ┆ f64         ┆ f64          ┆ f64         ┆ str            │
╞════════════╪══════════════╪═════════════╪══════════════╪═════════════╪════════════════╡
│ count      ┆ 150.0        ┆ 150.0       ┆ 150.0        ┆ 150.0       ┆ 150            │
│ null_count ┆ 0.0          ┆ 0.0         ┆ 0.0          ┆ 0.0         ┆ 0              │
│ mean       ┆ 5.843333     ┆ 3.054       ┆ 3.758667     ┆ 1.198667    ┆ null           │
│ std        ┆ 0.828066     ┆ 0.433594    ┆ 1.76442      ┆ 0.763161    ┆ null           │
│ min        ┆ 4.3          ┆ 2.0         ┆ 1.0          ┆ 0.1         ┆ Iris-setosa    │
│ 25%        ┆ 5.1          ┆ 2.8         ┆ 1.6          ┆ 0.3         ┆ null           │
│ 50%        ┆ 5.8          ┆ 3.0         ┆ 4.4          ┆ 1.3         ┆ null           │
│ 75%        ┆ 6.4          ┆ 3.3         ┆ 5.1          ┆ 1.8         ┆ null           │
│ max        ┆ 7.9          ┆ 4.4         ┆ 6.9          ┆ 2.5         ┆ Iris-virginica │
└────────────┴──────────────┴─────────────┴──────────────┴─────────────┴────────────────┘

As can be seen from the output above, there are **4 input features** and **1 class label** with a total of 150 rows, a very small dataset so it should be very easy to train. Fortunately, there are no missing values in the dataset hence we can use it as is.


In [5]:
data.group_by(["class"]).count()

shape: (3, 2)
┌─────────────────┬───────┐
│ class           ┆ count │
│ ---             ┆ ---   │
│ str             ┆ u32   │
╞═════════════════╪═══════╡
│ Iris-virginica  ┆ 50    │
│ Iris-versicolor ┆ 50    │
│ Iris-setosa     ┆ 50    │
└─────────────────┴───────┘

## Pre-processing

- Features are so small that we don't need to do feature selection
- Train test split / train-validate-test split
- Do we need Sampling?

## Ensemble

- Put all those models together

- cross fold validation?